src/compiler/checker.ts

```diff
@@ -28043,6 +28039,36 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getUnionType(map(clauseWitnesses, text => text ? narrowTypeByTypeName(type, text) : neverType));
         }
 
+        function narrowTypeBySwitchOnTrue(type: Type, switchStatement: SwitchStatement, clauseStart: number, clauseEnd: number): Type {
+            const defaultIndex = findIndex(switchStatement.caseBlock.clauses, clause => clause.kind === SyntaxKind.DefaultClause);

```

Comment:
The only time `switchStatement` is used is to get `caseBlock` which is only used for `clauses`. Maybe we just pass that in or destructure that out?

Reply:
I don't really know what `clauseStart === clauseEnd` means, but you could also avoid the `findIndex` by folding the walk into the first walk, and then you'd be able to avoid the range check, right?

You are basically walking across every `switch/case` per group of clauses (or per clause) which is N^2, right?

Reply:
Like the other stuff, this is just me doing what the other switch case narrowings do. I can do whatever but it'll look different when it's possible that all of them could be improved.

Reply:
Fair enough, we should come back to these at another PR.

Reply:
Yeah, having now touched this, I do think there's a bunch to improve here and elsewhere. It's extremely likely that this just hasn't been touched in a while (when do we ever add new narrowings???)

Reply:
Do you want me to leave a TODO behind?